### ISSUE TITLE
strictNullChecks still not supported in 4.1.2

### ISSUE THREAD

**Issue by lassekalheim** [angular/angular#16720]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
```

**Current behavior**
running `ng serve --prod` produces `Object is possibly 'null'` errors.

it also produces errors from angular code: 

`{{project path}}/node_modules/@angular/forms/src/model.d.ts (176,22): Type '{ onlySelf?: boolean | undefined; } | undefined' has no property 'onlySelf' and no string index signat
ure.`

**Expected behavior**
   compilation without error

If the current behavior is a bug or you can illustrate your feature request better with an example, 
please provide the *STEPS TO REPRODUCE* and if possible a *MINIMAL DEMO* of the problem via
https://plnkr.co or similar (you can use this template as a starting point: http://plnkr.co/edit/tpl:AvJOMERrnz94ekVua0u5).
-->

**What is the motivation / use case for changing the behavior?**

**Please tell us about your environment:**

Windows 10
* **Angular version:** 4.1.2

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

 
* **Language:** [TypeScript 2.3.2]

*** tsconfig ***
```json
{
  "compileOnSave": false,
  "compilerOptions": {
    "outDir": "./dist/out-tsc",
    "baseUrl": "src",
    "sourceMap": true,
    "declaration": false,
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "strictNullChecks": true,
    "noImplicitAny": true,
    "suppressImplicitAnyIndexErrors": true,
    "target": "es5",
    "typeRoots": [
      "node_modules/@types"
    ],
    "lib": [
      "es2016",
      "dom"
    ]
  }
}
```
* **Node (for AoT issues):** `node --version` =   7.5.0

**Comment by tbosch** [comment]
Yes, we had a problem with 4.1. 
I just tried with 4.5.0 and it works.

**Comment by tbosch** [comment]
Sorry, the version numbers are wrong.

**Comment by tbosch** [comment]
So I tried this locally and it works. Could you create a reproduction in a Github branch?

**Comment by mhevery** [comment]
This was fixed in 8c09d10ba9838216a01962615ddb3efa2106fcb6 and released in 4.2.0-beta.0 But somehow we forgot to release it in the patch branch.

**Comment by mhevery** [comment]
No I am mistaken, it was released in 4.1.1

**Comment by mhevery** [comment]
I just verified on 4.1.2 and I can't reproduce the issue. Closing. Reopen with better repro steps

**Comment by hill0826** [comment]
It looks like this might be an issue with not yet supporting TSLint 5+.

I'm able to reproduce this when using TSLint 5.2.0. I suspect the original poster was using TSLint 5+, too. Compilation succeeds when switching back to TSLint 4.5.1.

It doesn't look like Angular 4.1.2 is using TSLint 5.x.x yet, but it would be nice to have. Codelyzer 3.0.x does already have a dependency on TSLint 5.x.x.

Relevant environment:

- Angular: 4.1.2
- Angular CLI: 1.0.3
- TypeScript: 2.3.2
- Codelyzer: 3.0.1
- TSLint: 5.2.0

**Comment by hill0826** [comment]
I'm at a loss as to what's going on here. It's no longer showing up for me.

It's possible yarn left something stale in `node_modules` when I updated. That's the only thing I can think of for now.

This isn't blocking me from using TSLint 5+ or any of my other desired dependencies anymore. I'll update with a new comment if I run into the issue again.

**Comment by sfabriece** [comment]
I was definitely using tslint 5xx. I will check if I still run into this problem.

**Comment by Delagen** [comment]
this fragment of typings
```typescript
export interface ViewChildDecorator {
    /**
     * @whatItDoes Configures a view query.
     *
     * @howToUse
     *
     * {@example core/di/ts/viewChild/view_child_howto.ts region='HowTo'}
     *
     * @description
     *
     * You can use ViewChild to get the first element or the directive matching the selector from the
     * view DOM. If the view DOM changes, and a new child matches the selector,
     * the property will be updated.
     *
     * View queries are set before the `ngAfterViewInit` callback is called.
     *
     * **Metadata Properties**:
     *
     * * **selector** - the directive type or the name used for querying.
     * * **read** - read a different token from the queried elements.
     *
     * {@example core/di/ts/viewChild/view_child_example.ts region='Component'}
     *
     * **npm package**: `@angular/core`
     *
     * @stable
     * @Annotation
     */
    (selector: Type<any> | Function | string, {read}?: {
        read?: any;
    }): any;
    new (selector: Type<any> | Function | string, {read}?: {
        read?: any;
    }): ViewChild;
}
```
emit
```
TS2459: Type '{ read?: any; } | undefined' has no property 'read' and no string index signature.
```

and this
```typescript
export declare class RequestOptions {
    /**
     * Http method with which to execute a {@link Request}.
     * Acceptable methods are defined in the {@link RequestMethod} enum.
     */
    method: RequestMethod | string | null;
    /**
     * {@link Headers} to be attached to a {@link Request}.
     */
    headers: Headers | null;
    /**
     * Body to be used when creating a {@link Request}.
     */
    body: any;
    /**
     * Url with which to perform a {@link Request}.
     */
    url: string | null;
    /**
     * Search parameters to be included in a {@link Request}.
     */
    params: URLSearchParams;
    /**
     * @deprecated from 4.0.0. Use params instead.
     */
    /**
     * @deprecated from 4.0.0. Use params instead.
     */
    search: URLSearchParams;
    /**
     * Enable use credentials for a {@link Request}.
     */
    withCredentials: boolean | null;
    responseType: ResponseContentType | null;
    constructor({method, headers, body, url, search, params, withCredentials, responseType}?: RequestOptionsArgs);
    /**
     * Creates a copy of the `RequestOptions` instance, using the optional input as values to override
     * existing values. This method will not change the values of the instance on which it is being
     * called.
     *
     * Note that `headers` and `search` will override existing values completely if present in
     * the `options` object. If these values should be merged, it should be done prior to calling
     * `merge` on the `RequestOptions` instance.
     *
     * ```typescript
     * import {RequestOptions, Request, RequestMethod} from '@angular/http';
     *
     * const options = new RequestOptions({
     *   method: RequestMethod.Post
     * });
     * const req = new Request(options.merge({
     *   url: 'https://google.com'
     * }));
     * console.log('req.method:', RequestMethod[req.method]); // Post
     * console.log('options.url:', options.url); // null
     * console.log('req.url:', req.url); // https://google.com
     * ```
     */
    merge(options?: RequestOptionsArgs): RequestOptions;
    private _mergeSearchParams(params?);
    private _parseParams(objParams?);
    private _appendParam(key, value, params);
}
```
emit
```
TS2459: Type 'ResponseOptionsArgs | undefined' has no property '***' and no string index signature.
```
for every property from list of constructor arguments (as properties of object)
the same for `ResponseOptions`
